components/places/src/storage/bookmarks.rs

```diff
@@ -883,21 +814,79 @@ pub fn fetch_tree(db: &impl ConnExt, item_guid: &SyncGuid) -> Result<Option<Book
 
     let mut stmt = db.conn().prepare(sql)?;
 
-    let results =
+    let mut results =
         stmt.query_and_then_named(&[(":item_guid", item_guid)], FetchedTreeRow::from_row)?;
-    // markh can't work out how to do this directly with an AndThenRows - so
-    // we slurp the rows into a Vec and iterate over that.
-    let mut rows = Vec::new();
+
+    // The first row in the result set is always the root of our tree.
+    let mut root = match results.next() {
+        Some(result) => {
+            let row = result?;
+            FolderNode {
+                guid: Some(row.guid.clone()),
+                date_added: Some(row.date_added),
+                last_modified: Some(row.last_modified),
+                title: row.title.clone(),
+                children: Vec::new(),
+            }
+            .into()
+        }
+        None => return Ok(None),
+    };
+
+    // For all remaining rows, build a pseudo-tree that maps parent GUIDs to
```

Comment:
@thomcc, could you double-check 👀 this, please?

Reply:
`HashMap<Option<SyncGuid>, Vec<BookmarkTreeNode>>` huh that's a bit weird. Guess it makes some sense though.